File: Monotonic Stack/Q2. Daily Temperatures/dailyTemperatures.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temp) {
        int sz = temp.size();
        vector<int> ans(sz, 0);

        //Monotonic Tools
        stack<int> stk;
        int pt = sz-1;

        while(pt > -1){
            if(stk.empty()){
                stk.push(pt);
                pt--;
            }
            else{
                if(temp[stk.top()] > temp[pt]){
                    ans[pt] = stk.top() - pt;
                    stk.push(pt);
                    pt--;
                }
                else{
                    stk.pop();
                }
            }
        }

        return ans;
    }
};

File: Monotonic Stack/Q2. Daily Temperatures/dailyTemperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function (temp) {

    const len = temp.length
    let ans = Array(len).fill(0)
    let pt = len - 1
    let stk = []

    //O(n^2) Nested Loop
    // for(let i=0; i<len-1; i++){
    //     for(let j=i+1; j<len; j++){
    //         if(temp[j] > temp[i]){
    //             ans[i] = j-i
    //             break
    //         }
    //     }
    // }

    //O(n) Monotonic Stack
    while (pt > -1) {
        //if stack not empty
        if (stk.length) {
            if (temp[stk[stk.length - 1]] > temp[pt]) {
                ans[pt] = stk[stk.length - 1] - pt
                stk.push(pt)
                pt--
            }
            else {
                stk.pop()
            }
        }
        //Stack is empty
        else {
            stk.push(pt)
            pt--
        }

    }

    return ans
};